### ISSUE TITLE
nix-shell breaks if bash is in scope

### ISSUE THREAD

**Issue by bgamari** [NixOS/nixpkgs#29960]
## Issue description

Terminal handling within `nix-shell` appears to break if the `bash` derivation is in scope.

### Steps to reproduce

First verify that `nix-shell` works if `bash` isn't in scope,
```
[ben@nixos-unstable:~]$ nix-env -e bash
uninstalling ‘bash-4.4-p12’
[ben@nixos-unstable:~]$ nix-shell '<nixpkgs>' -p rr

[nix-shell:~]$ 
```
Try pressing up arrow to go back in shell history, note that it works as expected. Send mouse event to TTY (e.g. click in a mouse-capable terminal emulator), note that nothing happens.

Now try bringing `bash` into scope,
```
[ben@nixos-unstable:~]$ nix-env -i bash
installing ‘bash-4.4-p12’
[ben@nixos-unstable:~]$ nix-shell '<nixpkgs>' -p rr

\[\][nix-shell:~]$\[\] 
```
Note how prompt is mangled, pressing up arrow key results in `^[[A` being echoed to terminal instead of moving within shell history. Things are generally broken.

## Technical details

* System: `18.03pre116572.9824ca6975 (Impala)`, also confirmed on `17.03`.
* Nix version: `nix-env (Nix) 1.11.15`
* Nixpkgs version: `"18.03pre116572.9824ca6975"`
* Sandboxing enabled: `build-use-sandbox = false`
* `nix-shell` version: `nix-shell (Nix) 1.11.15`

**Comment by bgamari** [comment]
@cleverca22 points out on `#nixos` that `bash` lacks interactive support. I should really be using `bash-interactive`.

**Comment by danbst** [comment]
I observe same "picture" when run `nix-shell` inside `nix-shell`, btw

```
[danbst@LX:~]$ nix-shell -p stdenv

[nix-shell:~]$ nix-shell -p stdenv

\[\][nix-shell:~]$\[\] 
```

**Comment by yegortimoshenko** [comment]
This is a [Nix](https://github.com/NixOS/nix) issue.

**Comment by yegortimoshenko** [comment]
And I can't reproduce it with `nixUnstable`:

```
[chronos@yegatool:/tmp]$ nix-env -iA nixos.nixUnstable
installing ‘nix-unstable-1.12pre5732_fd10f6f2’
building path(s) ‘/nix/store/wkbfp2anpzcv6b16mr7h97vmr7nqjvs5-user-environment’
created 7 symlinks in user environment

[chronos@yegatool:/tmp]$ nix-shell --version
nix-shell (Nix) 1.12pre5732_fd10f6f2

[chronos@yegatool:/tmp]$ nix-env -i bash
installing 'bash-4.4-p12'

[chronos@yegatool:/tmp]$ nix-shell '<nixpkgs> -p rr
>
```

```
[chronos@yegatool:/tmp]$ nix-env -iA nixos.nixUnstable
installing ‘nix-unstable-1.12pre5732_fd10f6f2’
building path(s) ‘/nix/store/wkbfp2anpzcv6b16mr7h97vmr7nqjvs5-user-environment’
created 7 symlinks in user environment

[chronos@yegatool:/tmp]$ which nix-shell
/home/chronos/.nix-profile/bin/nix-shell

[chronos@yegatool:/tmp]$ nix-shell --version
nix-shell (Nix) 1.12pre5732_fd10f6f2

[chronos@yegatool:/tmp]$ nix-shell -p stdenv

[nix-shell:/tmp]$ nix-shell -p stdenv

[nix-shell:/tmp]$ nix-shell -p stdenv

[nix-shell:/tmp]$ 
```

**Comment by danbst** [comment]
Yeah, looks like https://github.com/NixOS/nix/commit/c94f3d5575d7af5403274d1e9e2f3c9d72989751 fixed it

**Comment by danbst** [comment]
Actually, this is quite a duplicate
https://github.com/NixOS/nixpkgs/issues/27126
https://github.com/NixOS/nix/issues/756
https://github.com/NixOS/nix/issues/730

And pending fix for 1.11.x
https://github.com/NixOS/nix/pull/1483

**Comment by jcrben** [comment]
The PR pointed to above https://github.com/NixOS/nix/pull/1483 says that it's a backport...

On Darwin, I still get a broken shell if I run a pure shell (arrow buttons do not work):
![image](https://user-images.githubusercontent.com/5614134/39670711-4a67cc8c-50bf-11e8-9014-03091675ebe1.png)

Passing `-p bashInteractive` fixes the shell, but best if it's unbroken by default:
![image](https://user-images.githubusercontent.com/5614134/39670716-7a089660-50bf-11e8-91f6-da827b8be61a.png)

But on my NixOS 18.03 Thinkpad, not having the same problem.

**Comment by qweered** [comment]
Doesn't reproduce on current Nix/nixpkgs: `nix-shell -p bash` keeps the running shell as `bashInteractive` (with readline) and only adds `pkgs.bash` to PATH, so the broken-readline / `^[[A` symptom from 2017 is gone. Closing as stale.
